File: LC/2976_Minimum_Cost_to_Convert_String_I/solution.py
class Solution:
    def minimumCost(self, source: str, target: str, original: List[str], changed: List[str], cost: List[int]) -> int:
        cost_graph = [[inf for _ in range(26)] for _ in range(26)]

        for ori, chg, c in zip(original, changed, cost):
            u = ord(ori) - ord('a')
            v = ord(chg) - ord('a')
            cost_graph[u][v] = min(cost_graph[u][v], c)

        for k in range(26):
            for i in range(26):
                for j in range(26):
                    cost_graph[i][j] = min(cost_graph[i][j], cost_graph[i][k] + cost_graph[k][j])

        total_cost = 0
        for i in range(len(source)):
            if source[i] == target[i]:
                continue

            r = ord(source[i]) - ord('a')
            c = ord(target[i]) - ord('a')
            
            if cost_graph[r][c] == inf:
                return -1
            
            total_cost += cost_graph[r][c]

        return total_cost

File: LC/2976_Minimum_Cost_to_Convert_String_I/solution.cpp
class Solution {
public:
    long long minimumCost(string source, string target, vector<char>& original, vector<char>& changed, vector<int>& cost) {
        vector<vector<long long>> cost_graph(26, vector<long long>(26, INT_MAX));

        for (int i = 0; i < original.size(); i++) {
            int u = (int)original[i] - (int)'a';
            int v = (int)changed[i] - (int)'a';
            cost_graph[u][v] = min(cost_graph[u][v], (long long)cost[i]);
        }

        for (int k = 0; k < 26; k++) {
            for (int i = 0; i < 26; i++) {
                for (int j = 0; j < 26; j++) {
                    cost_graph[i][j] = min(cost_graph[i][j], cost_graph[i][k] + cost_graph[k][j]);
                }
            }
        }

        long long total_cost = 0;
        for (int i = 0; i < source.length(); i++) {
            if (source[i] == target[i]) {
                continue;
            }

            int r = (int)source[i] - (int)'a';
            int c = (int)target[i] - (int)'a';

            if (cost_graph[r][c] == (long long)INT_MAX) {
                return -1;
            }

            total_cost += cost_graph[r][c];
        }

        return total_cost;
    }
};